Use vectorized load/store in GatherOpKernel

benbarsdell opened the pull request:
- Vectorized load/store achieves higher memory throughput.
- This commit chooses the optimal vector size that is compatible with the data shape and pointer alignment.
- Note that the `ldg` function would require more changes to support the vectorized types, but it is not actually needed: the use of `const __restrict__` means the kernel still generates `LDG` instructions without it.
- Some quick benchmarking showed the following speedups with this commit when gathering vectors of length 256 on an NVIDIA TITAN V GPU:
    float64: 1.02x
    float32: 1.07x
    float16: 1.35x
    bool:    1.67x

cc @nluehr

benbarsdell committed:
Use vectorized load/store in GatherOpKernel

- Vectorized load/store achieves higher memory throughput.
- This commit chooses the optimal vector size that is compatible with
  the data shape and pointer alignment.
- Note that the ldg function would require more changes to support the
  vectorized types, but it is not actually needed: the use of 'const
  __restrict__' means the kernel still generates LDG instructions
  without it.
- Some quick benchmarking showed the following speedups with this commit
  when gathering vectors of length 256 on an NVIDIA TITAN V GPU:
    float64: 1.02x
    float32: 1.07x
    float16: 1.35x
    bool:    1.67x

benbarsdell committed:
Fix alignment_of(T*) to return elements not bytes

benbarsdell committed:
Fix AlignedVector uniform value constructor

- The old constructor didn't work as expected because Eigen::half has a
  template constructor that accepts anything (which means
  std::is_constructible always returns true) but its implementation can
  still fail for some types.

benbarsdell committed:
Refactor vectorization dispatch in gather kernels

- This is cleaner and allows other kernels to use DispatchToVectorized
  too in future.

benbarsdell commented:
I've just pushed a couple of fixes for issues I noticed since submitting, and also a small refactor to make it cleaner. The commit messages have the details.

sanjoy commented on tensorflow/core/kernels/gather_functor_batched_gpu.cu.h:
```diff
@@ -29,10 +29,11 @@ namespace tensorflow {
-          bool is_axis_zero, bool is_batch_dims_zero>
-__global__ void GatherOpKernel(const T* params, const Index* indices, T* out,
-                               int64 outer_size,
+template <typename T, typename Index, bool is_axis_zero,
```
Can you please rename `T` to make it more obvious that it isn't necessarily a primitive type?

sanjoy commented on tensorflow/core/util/gpu_kernel_helper.h:
```diff
@@ -196,6 +198,87 @@ __device__ EIGEN_ALWAYS_INLINE Eigen::half GpuShuffleXorSync(
+inline int64 alignment_of(T* ptr) {
+  const intptr_t ptr_val = reinterpret_cast<std::uintptr_t>(ptr);
+  // Pointers should always be aligned to sizeof(T) bytes.
+  CHECK(ptr_val % sizeof(T) == 0);  // CRASH OK
```
`CHECK_EQ`.

sanjoy commented on tensorflow/core/util/gpu_kernel_helper.h:
```diff
@@ -196,6 +198,87 @@ __device__ EIGEN_ALWAYS_INLINE Eigen::half GpuShuffleXorSync(
+inline int64 alignment_of(int64 element_stride) {
+  return element_stride & -element_stride;
+}
+template <typename T>
```
Newline.

sanjoy commented on tensorflow/core/kernels/gather_functor_gpu.cu.h:
```diff
@@ -66,11 +66,56 @@ __global__ void GatherOpKernel(const T* __restrict__ params,
+                          int64 out_size) {
+  return DispatchToVectorized<
+      T, detail::LaunchGatherKernelVectorized<is_axis_zero>::template Impl>(
+      MinAlignmentOf(params, out, slice_size), d, params, indices, out,
```
This can be a bit confusing since the user will get vectorized operations or not depending on memory allocation patterns.  Do you have future plans to make this more robust around memory alignment (for instance by using non-vector loads/stores for the unaligned prefix)?  Or do you think it doesn't matter in practice?

benbarsdell replied:
I believe the allocator always returns aligned pointers, so I don't expect the alignment to vary.
In addition, doing non-vectorized ld/st for the prefix is not practical for multi-dimensional operations with a non-aligned stride because each row has a different alignment.

sanjoy replied:
Great, can you please add a comment that states this?

benbarsdell replied:
Added.

benbarsdell committed:
Address minor review comments from PR 46842

- Use CHECK_EQ instead of CHECK.
- Rename template arg T to ValueOrVec to indicate that it may not be
  a primitive type.
- Add a newline between function definitions.

benbarsdell committed:
Add comments re alignment in gather op kernels

benbarsdell commented:
Is something blocking this from being merged?

sanjoy commented:
> Is something blocking this from being merged?

Super annoying reason: the "Crask OK" annotations you added are not on the right lines.  For now I'll manually land the PR after changing the CHECKs to DCHECKs, I hope that's OK.

benbarsdell commented:
Sounds good, thanks.

mihaimaruseac commented:
This has been rolled back as it uses C++17 features. TF only supports C+14 at the moment.

benbarsdell commented:
What C++17 features? It successfully built with C++14 locally and in the github CI.

benbarsdell commented:
Oh I think I found it: clang complains about `template <int vec_size> typename Functor` instead of `template <int vec_size> class Functor`. (gcc does not complain).
This is a trivial fix. What's the best way to make it?

sanjoy commented:
@benbarsdell I rolled this forward with a fix.
